kubernetes-sigs/kubebuilder
# Using MonitorService to setup Prometheus

**Description of the change:**
Scaffold out optional ServiceMonitor objects to setup monitor metrics for Prometheus

**Motivation for the change:**
SDK Integration:
- https://github.com/kubernetes-sigs/kubebuilder/issues/887
Prometheus upstream no longer recommends using annotations.

**Following the steps to test it**
1. Install https://github.com/coreos/kube-prometheus
2. With this PR create a project to test it following the quick start

```shell
$ mkdir monitor
$ cd monitor/
$ kubebuilder init --domain monitor.domain
$ kubebuilder create api --group example --version v1 --kind MonitorTest
```

3. Uncomment the  ../prometheus in the `config/default/kustomization.yaml` to apply the ServiceMonitor 
4. Then, run the commands to apply all and build the project and deploy it.

```
$ make install
$ kubectl apply -f config/samples/
$ make docker-build docker-push IMG=cmacedo/monitor:0.0.1 # use  your registry
$ make deploy IMG=cmacedo/monitor:0.0.1
``` 

5. Check the resources it should have a metric service and the ServiceMonitor. 

* Check all resources

```
$ kubectl get  all -n monitor-system
NAME                                              READY   STATUS    RESTARTS   AGE
pod/monitor-controller-manager-7488d8fd54-s6gt6   2/2     Running   0          33s

NAME                                                 TYPE        CLUSTER-IP      EXTERNAL-IP   PORT(S)    AGE
service/monitor-controller-manager-metrics-service   ClusterIP   10.99.246.247   <none>        8443/TCP   33s

NAME                                         READY   UP-TO-DATE   AVAILABLE   AGE
deployment.apps/monitor-controller-manager   1/1     1            1           33s

NAME                                                    DESIRED   CURRENT   READY   AGE
replicaset.apps/monitor-controller-manager-7488d8fd54 
```
* Check the ServiceMonitor

```
$ kubectl get ServiceMonitor -n monitor-system 
NAME                                         AGE
monitor-controller-manager-metrics-monitor   2m8s
```
* Check the ServiceMonitor values

```
$ kubectl describe ServiceMonitor/monitor-controller-manager-metrics-monitor -n monitor-system
Name:         monitor-controller-manager-metrics-monitor
Namespace:    monitor-system
Labels:       control-plane=controller-manager
Annotations:  kubectl.kubernetes.io/last-applied-configuration:
                {"apiVersion":"monitoring.coreos.com/v1","kind":"ServiceMonitor","metadata":{"annotations":{},"labels":{"control-plane":"controller-manage...
API Version:  monitoring.coreos.com/v1
Kind:         ServiceMonitor
Metadata:
  Creation Timestamp:  2019-10-02T09:50:53Z
  Generation:          1
  Resource Version:    26878
  Self Link:           /apis/monitoring.coreos.com/v1/namespaces/monitor-system/servicemonitors/monitor-controller-manager-metrics-monitor
  UID:                 b30307ca-bdb7-47b1-adab-29e5ecb7b690
Spec:
  Endpoints:
    Path:  /metrics
    Port:  https
  Selector:
    Control - Plane:  controller-manager
Events:               <none>
```
* Check the Service values

```service
$ kubectl describe service/monitor-controller-manager-metrics-service -n monitor-system
Name:              monitor-controller-manager-metrics-service
Namespace:         monitor-system
Labels:            control-plane=controller-manager
Annotations:       kubectl.kubernetes.io/last-applied-configuration:
                     {"apiVersion":"v1","kind":"Service","metadata":{"annotations":{},"labels":{"control-plane":"controller-manager"},"name":"monitor-controlle...
Selector:          control-plane=controller-manager
Type:              ClusterIP
IP:                10.99.246.247
Port:              https  8443/TCP
TargetPort:        https/TCP
Endpoints:         172.17.0.9:8443
Session Affinity:  None
Events:            <none>
```

* Search for the metrics in the namespace where the POC was deployed. 

Search:  `{namespace="monitor-system"}`

<img width="1680" alt="Screenshot 2019-10-02 at 13 07 13" src="https://user-images.githubusercontent.com/7708031/66042888-a497da80-e515-11e9-9d77-d8a9fc1159a5.png">
